sql/parser/expr.go

```diff
@@ -286,11 +286,27 @@ func (p *Parser) parseType() document.ValueType {
 		return document.BlobValue
 	case scanner.TYPEDOCUMENT:
 		return document.DocumentValue
+	case scanner.REAL:
+		return document.DoubleValue
 	case scanner.TYPEDOUBLE:
+		tok, _, _ := p.ScanIgnoreWhitespace()
+		if tok == scanner.DOUBLEPRECISION {
+			return document.DoubleValue
+		}
+		p.Unscan()
 		return document.DoubleValue
-	case scanner.TYPEINTEGER:
+	case scanner.TYPEINTEGER, scanner.INT, scanner.INT2, scanner.INT8, scanner.TINYINT,
+		 scanner.BIGINT, scanner.MEDIUMINT, scanner.SMALLINT:
 		return document.IntegerValue
 	case scanner.TYPETEXT:
+		return document.TextValue
+	case scanner.VARCHAR, scanner.CHARACTER:
+		// The value between parentheses is not used.
+		_, err := p.parseExprList(scanner.LPAREN, scanner.RPAREN)
```

Comment:
This would make the following valid:

```SQL
VARCHAR(1 IN [1, 2, 3] AND foo > 4)
```

I understand the trick but parsers should be very strict.
We should expect one left parenthesis, an integer, and a closing parenthesis.